### PR TITLE
[CVP][LVI] Add support for vectors

### PR DESCRIPTION
**Pull request by nikic**
The core change here is to add support for converting vector constants into constant ranges. The rest is just relaxing isIntegerTy() checks and making sure we don't use APIs that assume vectors.

There are a couple of places that don't support vectors yet, most notably the "simplest" fold (comparisons to a constant) isn't supported yet. I'll leave these to a followup.

### CODE REVIEW

File: llvm/lib/Analysis/LazyValueInfo.cpp [hunk 1]
```diff
@@ -844,6 +862,9 @@ static ConstantRange toConstantRange(const ValueLatticeElement &Val,
   unsigned BW = Ty->getScalarSizeInBits();
   if (Val.isUnknown())
     return ConstantRange::getEmpty(BW);
+  if (Val.isConstant())
+    if (auto *VTy = dyn_cast<FixedVectorType>(Ty))
+      return getConstantRangeFromVector(Val.getConstant(), VTy);
```
[the remark this answers is not in the document]
Comment (nikic) [reply llvm/lib/Analysis/LazyValueInfo.cpp]: Good point. I've added scalable vectors support.

[the remark this answers is not in the document]
Comment (goldsteinn) [reply llvm/lib/Analysis/LazyValueInfo.cpp]: What about just using `computeConstantRange` to do it?
AFAICT the only case they handle differently is `poison` elements.

[the remark this answers is not in the document]
Comment (nikic) [reply llvm/lib/Analysis/LazyValueInfo.cpp]: I don't want to use computeConstantRange() here, but extracting a common helper would make sense as a followup, as part of making computeConstantRange() handle poison elements.

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Comment by dyung** [comment]
@nikic we are seeing an assertion failure in a test which I bisected back to this change. I have put the details in #97674, can you take a look?

**Comment by goldsteinn** [comment]
> @nikic we are seeing an assertion failure in a test which I bisected back to this change. I have put the details in #97674, can you take a look?

Fix at: https://github.com/llvm/llvm-project/pull/97682

File: llvm/lib/Analysis/LazyValueInfo.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (goldsteinn) [reply llvm/lib/Analysis/LazyValueInfo.cpp]: Ah looks like you beat me too it

**Pull request merged**